### PR TITLE
Add Paging Interface for LDAP Connection

### LINKED ISSUE #7702: LDAP search failed: Result Code 4 "Size Limit Exceeded"
**Describe the bug**
LDAP queries that return a large number of results are still erroring with the fix for #4162 . When an LDAP server is setup to only return 1000 results a group query which returns more than that will fail.

**To Reproduce**
1. Setup LDAP auth with a group filter that will return more than 1000 groups
2. Attempt to login
3. Login fails with `LDAP search failed: LDAP Result Code 4 "Size Limit Exceeded"`

**Expected behavior**
The login should be successful and all groups should be returned in the query response so authentication will succeed.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.2.3
* Vault CLI Version (retrieve with `vault version`): 1.2.3
* Server Operating System/Architecture: Linux

Vault server configuration file(s):

```
vault write auth/ldap/config \
  url="ldaps://ldap.example.com:636" \
  userdn="OU=Users,OU=Org,OU=BusinessUnits,DC=example,DC=com" \
  groupdn="OU=AccountGroups,OU=Groups,OU=Org,OU=BusinessUnits,DC=example,DC=com" \
  groupfilter="(&(objectClass=group))" \
  groupattr="memberOf" \
  upndomain="example.com" \
  insecure_tls=false \
  starttls=false

vault write auth/ldap/groups/developers policies=dev
```

**Additional context**
To see the problem, the LDAP server must be setup to return fewer groups than the number of groups which match the query.

This search appears to need updated: https://github.com/hashicorp/vault/blob/master/sdk/helper/ldaputil/client.go#L120-L125

Should use something like this:

```
result, err := conn.SearchWithPaging(&ldap.SearchRequest{
  BaseDN:    cfg.UserDN,
  Scope:     ldap.ScopeWholeSubtree,
  Filter:    filter,
  SizeLimit: math.MaxInt32,
}, math.MaxInt32)
```

Example using go-ldap. Given the proper group query, the first search will fail and the second will succeed:

```
package main

import (
	"fmt"
	"log"
	"math"

	"github.com/go-ldap/ldap"
)

func main() {
	l, err := ldap.Dial("tcp", "ldap.example.com:389")
	if err != nil {
		log.Fatal(err)
	}
	defer l.Close()
	l.Bind("<binduser>", "<bindpw>")
	request := &ldap.SearchRequest{
		BaseDN: "OU=AccountGroups,OU=Groups,OU=Org,OU=BusinessUnits,DC=example,DC=com",
		Scope:  ldap.ScopeWholeSubtree,
		Filter: "(&(objectClass=group))",
		Attributes: []string{
			"member:1.2.840.113556.1.4.1941:",
		},
		SizeLimit: math.MaxInt32,
	}

	//ERROR: SIZE LIMIT EXCEEDED
	sr, err := l.Search(request)
	if err != nil {
		log.Print(err)
	} else {
		for _, entry := range sr.Entries {
			fmt.Printf("%s: %v\n", entry.DN, entry.GetAttributeValue("cn"))
		}
	}

	//WORKS
	sr, err = l.SearchWithPaging(request, math.MaxInt32)
	if err != nil {
		log.Fatal(err)
	}

	for _, entry := range sr.Entries {
		fmt.Printf("%s: %v\n", entry.DN, entry.GetAttributeValue("cn"))
	}
}
```

### PR DESCRIPTION
**Pull request by ltcarbonell**
Vault was returning the following error when trying to filter on a large number of groups
```
* LDAP search failed: LDAP Result Code 4 "Size Limit Exceeded": 
```
This change makes use of paginated ldap searching, through a new connection interface as to maintain backwards compatibility for the SDK.

Fixes #7702

### CODE REVIEW

**Review by jasonodonnell — APPROVED** [review]
LGTM!

**Review by austingebauer — APPROVED** [review]
LGTM! 

Curious if you were able to test this out with group count > the MaxInt32 page size? So many groups 😅

File: sdk/helper/ldaputil/connection.go [hunk 1]
```diff
@@ -20,3 +20,8 @@ type Connection interface {
 	SetTimeout(timeout time.Duration)
 	UnauthenticatedBind(username string) error
 }
+
+type ConnectionPaging interface {
```
Comment (austingebauer) [inline sdk/helper/ldaputil/connection.go:24]: Nit: The other way around (`PagingConnection`) makes more sense as an interface name imo.

**Comment by ltcarbonell** [comment]
> LGTM!
> 
> Curious if you were able to test this out with group count > the MaxInt32 page size? So many groups 😅

@austingebauer unfortunately no, my machine wasn't a fan of that many groups 😆.  I did set it arbitrarily low to make sure that it would work if groups were larger than the page size.

**Review by fairclothjm — APPROVED** [review]

File: sdk/helper/ldaputil/connection.go (hunk 1 above)
Comment (ltcarbonell) [reply sdk/helper/ldaputil/connection.go:24]: I like that change, updated the name

File: sdk/helper/ldaputil/client.go [hunk 2]
```diff
@@ -351,6 +351,64 @@ func (c *Client) performLdapFilterGroupsSearch(cfg *ConfigEntry, conn Connection
...
+	// returning the user's groups
+	if c.Logger.IsDebug() {
+		c.Logger.Debug("compiling group filter", "group_filter", cfg.GroupFilter)
+	}
+
+	// Parse the configuration as a template.
+	// Example template "(&(objectClass=group)(member:1.2.840.113556.1.4.1941:={{.UserDN}}))"
+	t, err := template.New("queryTemplate").Parse(cfg.GroupFilter)
+	if err != nil {
+		return nil, errwrap.Wrapf("LDAP search failed due to template compilation error: {{err}}", err)
```
[the remark this answers is not in the document]
Comment (ltcarbonell) [reply sdk/helper/ldaputil/client.go]: Good idea.  Updated the references to that

**Pull request merged**